network/tls.go

```diff
@@ -0,0 +1,415 @@
+package network
+
+import (
+	"bytes"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/asn1"
+	"errors"
+	"math/big"
+	"net"
+	"time"
+
+	"github.com/dedis/kyber/sign/schnorr"
+	"github.com/dedis/kyber/util/encoding"
+	"github.com/dedis/kyber/util/random"
+	"github.com/dedis/onet/log"
+)
+
+// About our TLS strategy:
+//
+// The design goals were:
+// 1. use vanilla TLS for conode/conode communication, in order to ease deployment
+//    anxiety with industrial partners. ("If it's not TLS, it's not allowed." Sigh.)
+// 2. zero config: whatever config we have in an existing private.toml should be enough
+// 3. mutual authentication: each side should end up with proof that the other side is
+//    holding the secret key associated with the public key they claim to have.
+//
+// In order to achieve #1, we limit ourselves to TLS 1.2. This means that the TLS
+// private key must be different than the conode private key (which may be from a suite
+// not supported in TLS 1.2).
+//
+// In order to achieve #2, we use self-signed TLS certificates, with a private key
+// that is created on server boot and stored in RAM. Because the certificates are
+// self-signed, there is no need to pay a CA, nor to load the private key
+// or certificate from disk.
+//
+// In order to achieve #3, we include an extension in the certificate, which
+// proves that the same entity that holds the TLS private key (i.e. the signing
+// key for the self-signed TLS cert) also holds the conode's private key.
+// We do this by hashing a nonce provided by the peer (in order to prove that this
+// is a fresh challenge response) and the ASN.1 encoded CommonName of the certificate.
+// The CN is always the hex-encoded form of the conode's public key.
```

Comment:
does this mean we cannot use TLS in the normal way? i.e. proof that we actually own the domain name

Reply:
Yes, for conode/conode communication, our definition of "a valid certificate: is nonstandard. We can use CN's related to our domain name on the websocket side.